### PR TITLE
Add canned policy support

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
Add canned policy support

## Motivation and Context
This PR adds an additional API where we can create
a new set of canned policies which can be used with one
or many users.

## Regression
No

## How Has This Been Tested?
Using `mc admin policies`  from https://github.com/minio/mc/pull/2554
```
$ mc admin policies list myminio
everything          

$  mc admin policies list myminio everything
{"Version":"2012-10-17","Statement":[{"Effect":"Allow","Action":["s3:GetBucketLocation","s3:ListBucketMultipartUploads"],"Resource":["arn:aws:s3:::*"]},{"Effect":"Allow","Action":["s3:AbortMultipartUpload","s3:DeleteObject","s3:ListMultipartUploadParts","s3:PutObject"],"Resource":["arn:aws:s3:::*/*"]}]}
```

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [ ] I have added unit tests to cover my changes.
- [ ] I have added/updated functional tests in [mint](https://github.com/minio/mint). (If yes, add `mint` PR # here: )
- [x] All new and existing tests passed.

### CODE REVIEW

File: pkg/madmin/API.md [hunk 1]
```diff
@@ -38,10 +38,10 @@ func main() {
...
 |:----------------------------|:----------------------------|:--------------------------------------|:--------------------------|:------------------------------------|:------------------------------------|
-| [`ServiceStatus`](#ServiceStatus) | [`ServerInfo`](#ServerInfo) | [`Heal`](#Heal) | [`GetConfig`](#GetConfig) | [`AddUser()`](#AddUser) | [`SetAdminCredentials`](#SetAdminCredentials) |
+| [`ServiceStatus`](#ServiceStatus) | [`ServerInfo`](#ServerInfo) | [`Heal`](#Heal) | [`GetConfig`](#GetConfig) | [`AddUser`](#AddUser) | [`SetAdminCredentials`](#SetAdminCredentials) |
 | [`ServiceSendAction`](#ServiceSendAction) | | | [`SetConfig`](#SetConfig) | [`AddUserPolicy`](#AddUserPolicy) | [`StartProfiling`](#StartProfiling) |
 | | |            | [`GetConfigKeys`](#GetConfigKeys) | [`ListUsers`](#ListUsers) | [`DownloadProfilingData`](#DownloadProfilingData) |
-| | |            | [`SetConfigKeys`](#SetConfigKeys) |                                     |
+| | |            | [`SetConfigKeys`](#SetConfigKeys) | [`AddCannedPolicy`](#AddCannedPolicy) | |
```
Comment (poornas) [inline pkg/madmin/API.md:44]: RemoveCannedPolicy and ListCannedPolicies need to be added

**Review by poornas — CHANGES_REQUESTED** [review]
RemoveUserPolicy - this should be taking access key and policy name as argument - it doesnt actually do anything now. There should also be ability to list user policies.

File: pkg/madmin/API.md (hunk 1 above)
Comment (harshavardhana) [reply pkg/madmin/API.md:44]: Not added on purpose @poornas as it's not needed.. these APIs are low level and not meaningful.

**Comment by harshavardhana** [comment]
> RemoveUserPolicy - this should be taking access key and policy name as argument - it doesnt actually do anything now. There should also be ability to list user policies.

all of this is already there in `mc admin policies` - this is not going to be added in our code since its not necessary.

**Comment by harshavardhana** [comment]
> RemoveUserPolicy - this should be taking access key and policy name as argument - it doesnt actually do anything now. There should also be ability to list user policies.

A user only has only one policy so there is no reason to take policy name as an argument.

**Comment by harshavardhana** [comment]
> it doesnt actually do anything now. There should also be ability to list user policies.

Also added 

```
$ mc admin users list myminio
enabled    newuser               everything          
enabled    newuser1              everything     
```

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

File: cmd/admin-router.go [hunk 2]
```diff
@@ -81,15 +81,25 @@ func registerAdminRouter(router *mux.Router) {
...
+		Queries("accessKey", "{accessKey:.*}").Queries("name", "{name:.*}")
+
+	// Remove policy IAM
+	adminV1Router.Methods(http.MethodDelete).Path("/remove-canned-policy").HandlerFunc(httpTraceHdrs(adminAPI.RemoveCannedPolicy)).Queries("name", "{name:.*}")
 
 	// Remove user IAM
 	adminV1Router.Methods(http.MethodDelete).Path("/remove-user").HandlerFunc(httpTraceHdrs(adminAPI.RemoveUser)).Queries("accessKey", "{accessKey:.*}")
-	adminV1Router.Methods(http.MethodDelete).Path("/remove-user-policy").HandlerFunc(httpTraceHdrs(adminAPI.RemoveUserPolicy)).Queries("accessKey", "{accessKey:.*}")
+	adminV1Router.Methods(http.MethodDelete).Path("/remove-user-policy").HandlerFunc(httpTraceHdrs(adminAPI.RemoveUserPolicy)).
+		Queries("accessKey", "{accessKey:.*}").Queries("name", "{name:.*}")
```
Comment (poornas) [inline cmd/admin-router.go]: @harshavardhana, since there is only one policy / user - the policy name wouldn't need to be a query parameter.

Comment (harshavardhana) [reply cmd/admin-router.go]: Actually this API is not needed anymore.

File: cmd/admin-handlers.go [hunk 3]
```diff
@@ -899,39 +937,108 @@ func (a adminAPIHandlers) AddUserPolicy(w http.ResponseWriter, r *http.Request)
...
+		return
+	}
+
+	// Custom IAM policies not allowed for admin user.
+	if accessKey == globalServerConfig.GetCredential().AccessKey {
+		writeErrorResponseJSON(w, ErrInvalidRequest, r.URL)
+		return
+	}
+
+	if err := globalIAMSys.SetUserPolicy(accessKey, policyName); err != nil {
```
Comment (poornas) [inline cmd/admin-handlers.go:1001]: AddUserPolicy returns ok even if user is not defined yet - perhaps it should return appropriate error if user not present in globalIAMSys.

File: cmd/admin-handlers.go [hunk 4]
```diff
@@ -730,9 +730,13 @@ func (a adminAPIHandlers) RemoveUser(w http.ResponseWriter, r *http.Request) {
 	vars := mux.Vars(r)
 	accessKey := vars["accessKey"]
 	if err := globalIAMSys.DeleteUser(accessKey); err != nil {
-		logger.LogIf(ctx, err)
-		writeErrorResponseJSON(w, ErrInternalError, r.URL)
-		return
+		switch err.(type) {
+		case ObjectNotFound:
```
Comment (vadmeste) [inline cmd/admin-handlers.go]: It is not clear why we are avoiding returning an error when the user doesn't exist, can you add a comment for this ?

File: pkg/madmin/API.md [hunk 5]
```diff
@@ -362,15 +376,13 @@ __Example__
 ```
 
 <a name="AddUserPolicy"></a>
-### AddUserPolicy(user string, policy string) error
-Set a new policy for a given user on Minio server.
+### AddUserPolicy(user string, policyName string) error
```
Comment (vadmeste) [inline pkg/madmin/API.md]: If we can set only one canned policy per user, then let's call this function SetUserPolicy()

File: cmd/iam.go [hunk 6]
```diff
@@ -204,6 +219,44 @@ func (sys *IAMSys) DeletePolicy(accessKey string) error {
...
+	}
+
+	sys.Lock()
+	defer sys.Unlock()
+
+	p, ok := sys.iamCannedPolicyMap[policyName]
+	if !ok {
+		return errInvalidArgument
+	}
+	if p.IsEmpty() {
```
Comment (vadmeste) [inline cmd/iam.go]: I think this block is little confusing, if we are doing the check when adding canned policy, then this is not needed

File: cmd/iam.go [hunk 7]
```diff
@@ -332,8 +386,9 @@ func (sys *IAMSys) IsAllowed(args iampolicy.Args) bool {
 	}
 
 	// If policy is available for given user, check the policy.
-	if p, found := sys.iamPolicyMap[args.AccountName]; found {
-		return p.IsAllowed(args)
+	if name, found := sys.iamPolicyMap[args.AccountName]; found {
```
Comment (vadmeste) [inline cmd/iam.go:378]: Are we allowing by default if no policy is set ?

**Review by vadmeste — DISMISSED** [review]
Some comments

File: cmd/admin-handlers.go (hunk 4 above)
Comment (harshavardhana) [reply cmd/admin-handlers.go]: We are avoiding because its meaningless and not actionable to the client.

File: cmd/iam.go (hunk 7 above)
Comment (harshavardhana) [reply cmd/iam.go:378]: We don't this is no different than the previous code.

**Comment by minio-ops** [comment]
```
Error running mint automation
+ mkdir -p 6637-79c6661/gopath/src/github.com/minio
+ git clone --quiet https://github.com/minio/minio.git 6637-79c6661/gopath/src/github.com/minio/minio
++ cd 6637-79c6661/gopath/src/github.com/minio/minio
++ git remote add minio https://github.com/minio/minio.git
++ cd 6637-79c6661/gopath/src/github.com/minio/minio
++ git fetch --quiet minio pull/6637/head:pr6637
++ cd 6637-79c6661/gopath/src/github.com/minio/minio
++ git checkout --quiet pr6637
++ minikube ip
+ minikube_ip=192.168.39.64
++ mkdir -p certs
++ cd certs
++ ./cert -host 192.168.39.64
2018/10/16 12:18:22 wrote public.crt
2018/10/16 12:18:22 wrote private.key
+ GOPATH=/mnt/nvme0n1p1/mint-auto/mint-auto/6637-79c6661/gopath
+ make -C 6637-79c6661/gopath/src/github.com/minio/minio --quiet
# github.com/minio/minio/cmd
cmd/admin-router.go:89:99: adminAPI.AddUserPolicy undefined (type adminAPIHandlers has no field or method AddUserPolicy)
make: *** [build] Error 2

```

File: pkg/madmin/API.md (hunk 5 above)
Comment (harshavardhana) [reply pkg/madmin/API.md]: Done

File: cmd/admin-handlers.go (hunk 4 above)
Comment (harshavardhana) [reply cmd/admin-handlers.go]: Actually added an error anyways..

File: cmd/admin-handlers.go (hunk 3 above)
Comment (harshavardhana) [reply cmd/admin-handlers.go:1001]: Done

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :x: [more...](#0edfb48d-a437-43e6-be16-bd71b95a3a27) |
<a name="0edfb48d-a437-43e6-be16-bd71b95a3a27"></a>
### 6637-606eaaa/mint-gateway-nas.sh.log:
```
Running with
SERVER_ENDPOINT:      minikube:32689
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp aecc7647e58a:/mint/log /tmp/mint-logs'
(1/13) Running awscli tests ... done in 1 minutes and 13 seconds
(2/13) Running aws-sdk-go tests ... done in 0 seconds
(3/13) Running aws-sdk-java tests ... done in 4 seconds
(4/13) Running aws-sdk-php tests ... done in 41 seconds
(5/13) Running aws-sdk-ruby tests ... done in 5 seconds
(6/13) Running mc tests ... done in 38 seconds
(7/13) Running minio-dotnet tests ... FAILED in 2 minutes and 39 seconds

Unhandled Exception: System.AggregateException: One or more errors occurred. (Assert.Fail failed. ) ---> Microsoft.VisualStudio.TestTools.UnitTesting.AssertFailedException: Assert.Fail failed. 
   at Microsoft.VisualStudio.TestTools.UnitTesting.Assert.HandleFail(String assertionName, String message, Object[] parameters)
   at Minio.Functional.Tests.FunctionalTest.<>c.<ListIncompleteUpload_Test2>b__85_1(Exception ex) in /mint/run/core/minio-dotnet/FunctionalTest.cs:line 2138
   at System.Reactive.AutoDetachObserver`1.OnErrorCore(Exception exception) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Internal\AutoDetachObserver.cs:line 76
   at System.Reactive.AnonymousSafeObserver`1.OnError(Exception error) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\AnonymousSafeObserver.cs:line 62
   at System.Reactive.Linq.ObservableImpl.Throw`1._.Invoke() in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Linq\Observable\Throw.cs:line 42
   at System.Reactive.Concurrency.Scheduler.Invoke(IScheduler scheduler, Action action) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Concurrency\Scheduler.Simple.cs:line 83
   at System.Reactive.Producer`2.SubscribeRaw(IObserver`1 observer, Boolean enableSafeguard) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Internal\Producer.cs:line 127
   at System.Reactive.Linq.QueryLanguage.<>c__DisplayClass330_0`1.<Create>b__0(IObserver`1 observer) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Linq\QueryLanguage.Creation.cs:line 46
   at System.Reactive.AnonymousObservable`1.SubscribeCore(IObserver`1 observer) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\AnonymousObservable.cs:line 37
   at System.Reactive.ObservableBase`1.ScheduledSubscribe(IScheduler _, AutoDetachObserver`1 autoDetachObserver) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\ObservableBase.cs:line 82
   at System.Reactive.Concurrency.CurrentThreadScheduler.Trampoline.Run(SchedulerQueue`1 queue) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Concurrency\CurrentThreadScheduler.cs:line 126
   at System.Reactive.Concurrency.CurrentThreadScheduler.Schedule[TState](TState state, TimeSpan dueTime, Func`3 action) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\Concurrency\CurrentThreadScheduler.cs:line 94
   at System.Reactive.ObservableBase`1.Subscribe(IObserver`1 observer) in D:\a\1\s\Rx.NET\Source\src\System.Reactive\ObservableBase.cs:line 51
   at Minio.Functional.Tests.FunctionalTest.ListIncompleteUpload_Test2(MinioClient minio) in /mint/run/core/minio-dotnet/FunctionalTest.cs:line 2136
   --- End of inner exception stack trace ---
   at System.Threading.Tasks.Task.Wait(Int32 millisecondsTimeout, CancellationToken cancellationToken)
   at System.Threading.Tasks.Task.Wait()
   at Minio.Functional.Tests.FunctionalTest.Main(String[] args) in /mint/run/core/minio-dotnet/FunctionalTest.cs:line 241

Executed 6 out of 13 tests successfully.
```

File: pkg/madmin/user-commands.go [hunk 8]
```diff
@@ -137,43 +138,18 @@ func (adm *AdminClient) AddUser(accessKey, secretKey string) error {
...
-// AddUserPolicy - adds a policy for a user.
-func (adm *AdminClient) AddUserPolicy(accessKey, policy string) error {
+// SetUserPolicy - adds a policy for a user.
+func (adm *AdminClient) SetUserPolicy(accessKey, policyName string) error {
 	queryValues := url.Values{}
 	queryValues.Set("accessKey", accessKey)
+	queryValues.Set("name", policyName)
 
 	reqData := requestData{
 		relPath:     "/v1/add-user-policy",
```
Comment (vadmeste) [inline pkg/madmin/user-commands.go]: I think this needs to be also changed to set-user-policy

Comment (harshavardhana) [reply pkg/madmin/user-commands.go]: Done fixed

**Review by poornas — APPROVED** [review]
LGTM

**Review by vadmeste — APPROVED** [review]
LGTM & tested

**Comment by minio-ops** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-tls.sh | :heavy_check_mark: |
| mint-xl.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-worm.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-dist-xl.sh | :heavy_check_mark: |

**Pull request merged**